clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -1889,30 +1889,27 @@ void SYCLIntegrationHeader::emitFwdDecl(raw_ostream &O, const Decl *D,
     auto *NS = dyn_cast_or_null<NamespaceDecl>(DC);
 
     if (!NS) {
-      if (!DC->isTranslationUnit()) {
-        const TagDecl *TD = isa<ClassTemplateDecl>(D)
-                                ? cast<ClassTemplateDecl>(D)->getTemplatedDecl()
-                                : dyn_cast<TagDecl>(D);
-
-        if (TD && !UnnamedLambdaSupport) {
-          // defined class constituting the kernel name is not globally
-          // accessible - contradicts the spec
-          const bool KernelNameIsMissing = TD->getName().empty();
-          if (KernelNameIsMissing) {
+      const TagDecl *TD = isa<ClassTemplateDecl>(D)
+                              ? cast<ClassTemplateDecl>(D)->getTemplatedDecl()
+                              : dyn_cast<TagDecl>(D);
+      if (!TD)
+        break;
+
+      const bool KernelNameIsMissing = TD->getName().empty();
+      if (KernelNameIsMissing)
+        Diag.Report(KernelLocation, diag::err_sycl_kernel_incorrectly_named)
+            << /* kernel name is missing */ 0;
```

Comment:
I spend some time looking into these changes, and I don't get why we didn't start emitting "kernel name is missing" on each unnamed lambda, because before your changes this diagnostic was under `if (!UnnamedLambdaSupport)` and it looked correct.

Reply:
The original "if" was `if (TD && !UnnamedLambdaSupport)`, and when this evaluates false, then the else block would still use `TD` anyway -- which seems weird. premanandrao's commit refactored the TD check out to `if (!TD) break;`

Reply:
I'm still a bit confused about this change. Can you give an example where `if (TD)` evaluates true and KernelNameIsMissing i.e. TD does not have a name? 

Also, this change is not technically required for 'std' namespace check right? Or am I missing something?

Reply:
I think this happens when it is invalid.
It is not needed strictly for the specific 'std' namespace check, but it occurs in those contexts too.

Reply:
Normally the `if (TD)` check would ensure that a kernel name type is a TagDecl -- that is, one of a struct/union/class/enum -- which would catch kernel name types such as `int`. I say "normally", though, because earlier in the call stack, we also ensure that a kernel name type is a CXXRecordDecl -- that is, one of a struct/union/class. Which means, sadly, that the `if (TD)` part may actually be dead code. **But**, it was dead code even before, and independent of, this PR. I had considered doing a clean-up of the existing code, but... I feel like this PR has already exceed its side-quest quota. :-)

Reply:
Sorry if I'm being obtuse here but it still isn't clear to me what this diagnostic is trying to catch. Can you give an example of the invalid code this diagnostic triggers for. Maybe then I'll understand why this diagnostic no longer needs to be guarded by UnnamedLambdaSupport.

Reply:
There's still an `UnnamedLambdaSupport` guard, just moved a few lines down, currently at L#1905. In the original pre-PR code, this condition -- `if (TD && !UnnamedLambdaSupport) {` -- had a \*potential\* problem because the "if" proceeds if `TD` is truthy (non-null), yet the "else" block still uses `TD` anyway.

I wrote "potential" problem because it all appears to be dead code anyway. Up the call stack, in `SYCLIntegrationHeader::emit` where we're writing out the source of the header, if there's no UnnamedLambdaSupport, then we don't call into `emitFwdDecl` at all. Which means, in that original pre-PR condition, `TD` would always be true, and `!UnnamedLambdaSupport` would also always be true, making the entire "else" block -- and the diagnositic within -- dead code.

@premanandrao Is there a specific case that you were fixing? Or was this general tidying?

Reply:
I don't believe it is dead-code; at least it wasn't when I refactored it.   I am pretty sure one of the cases in the test (as it was originally) would fail if this part were reverted.   So there is a simple way to check if this is still needed.   (i.e. Revert the refactoring and see if the unnamed-kernel.cpp test still passes.)

Reply:
Locally, I reverted the refactoring and still got passes for all the CodeGenSYCL and SemaSYCL tests. The refactoring commit was from back in April, so yes certainly lots of changes to the codebase since then. :-)

Is that the decision, then? Do we want to revert that? We'll have to wait a few hours for buildbot tests and then we'll all need to re-approve again.

Reply:
Fixing that code to not have useless dead code would be my preference.

Reply:
That would be my preference too **\*\*EXCEPT\*\*** this PR has already exceed its side-quest quota. :-) This PR doesn't introduce the dead code, which means we can merge *this* PR then open a separate one to clean up any dead code if we still want to do that. So the options I'm personally considering are: revert or don't revert the original refactoring from April.

Reply:
May be I am misunderstanding or not seeing the diffs correctly, are we talking about the:
```
const TagDecl *TD = isa<ClassTemplateDecl>(D)
                              ? cast<ClassTemplateDecl>(D)->getTemplatedDecl()
                              : dyn_cast<TagDecl>(D);
      if (!TD)
        break;
```
change?   If so, isn't that part of what is being checked in?

Reply:
Yes, but the original had dead code too. This change just moved it to a different line.

Reply:
I see.  In that case, I don't mind this being addressed in a separate PR.

Reply:
OK then sounds like we're opting to *not* revert the April refactoring? And this PR would stay as is?

Reply:
Seems like folks still aren't comfortable with those changes, and since they come from an older time in this PR and don't seem to be needed anymore, I pushed a commit to revert that bit of refactoring. Buildbot is running. I'll ping folks when ready for re-review approvals.

Reply:
The changes @elizabethandrews was concerned about are reverted. All checks passed. Need review approvals again. @premanandrao @rbegam